### PR TITLE
Add tests, add docstrings, and fix small bugs

### LINKED ISSUE #17: Setup tests
It would be nice to have some primitive tests for this project.

### LINKED ISSUE #19: The `inherited-members` Sphinx directive option is respected backwards
The sphinx.ext.autodoc provided directives support a `inherited-members` option, and acts as an opt-in option - [defaulting to not including inherited members](https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#directive-option-automodule-members).

But I think we have implemented it backwards, so that we by default include inherited members, and when inherited-members is passed we don't include them.

A first step to address this issue could be to setup basic tests (#17), and/or investigate it further to verify how things work with and without the inherited-members directive option.

### LINKED ISSUE #26: `autoclass` is not the same as `autoconfigurable` - but should be?
I've realized that the `autoclass` directive is not the same as `autoconfigurable`.

1. We should update the README.md for now to not mention autoclass, but autoconfigurable.
2. We should consider providing a `can_document_member` function also for the TraitsletsDocumenter class, and perhaps increase the priority or similar. That is whats done for the other class for Attributes.

Unsure, but more investigations are needed.

### PR DESCRIPTION
**Pull request by consideRatio**
I've worked to understand, document, test, and fix smaller bugs I found.

### About tests

The tests/docs folder contains a sphinx project with examples used in the tests, which are overall simple should provide quite good coverage. They ensure that the sphinx directives are executed in various situations without warnings, and render .html strings of relevance.

It could have made sense to not render all the way to .html, but instead for example inspect the rST emitted by the directives somehow - but as I didn't knew how to do that I went for this end-to-end like test which probably is a good starting point to have for anyone looking to rework this further.

- Closes #17

### About small bugs

- I fixed a small edge case bug if a `:inherited-members:` options was passed to an `autoconfigurable` directive, but I don't think anyone use that as it clutters things.
  Closes #19
- The refactoring led to fixing a bug where `:exclude-members:` was bypassed, but will now be respected.
- I fixed a small bug in the README.md as introduced recently by me where I wrote about `autoclass` could be used instead of `autoconfigurable` - as it can't.
  Closes #26

### About documentation

Understanding autodoc_traits requires understanding of details of how sphinx.ext.autodoc works internally, which lacks documentation. Due to that I've invested a lot of time to help my future self and other contributors understand autodoc_traits by providing relevant references to sphinx source code and comments to summarize the important parts.

### Notes for review

I've probably invested ~16-24 hours into this. It took quite a while to onboard myself as I had to learn from comments in the sphinx source code to understand how things worked in more detail. 

I ended up adding multiple labels to this PR as it was quite extensive, which will make it show up in multiple categories if github-activity is used. A lot of the refactoring was very relevant documenting, but some refactoring also led to bugfixes etc.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -81,6 +83,14 @@ target_version = [
 ]
 
 
+# pytest is used for running Python based tests
+#
+# ref: https://docs.pytest.org/en/stable/
+#
+[tool.pytest.ini_options]
+addopts = "--verbose --color=yes --durations=10 --ignore=tests/docs/test_module.py"
```
Comment (minrk) [inline pyproject.toml]: Maybe it would be better to call it `sample_module` to avoid it being included as a test by default?

File: tests/docs/test_module.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+This module provides a module and classes with and without traits to test
+autodoc_traits against. It does not contain tests.
+"""
+
+from traitlets import Bool
+from traitlets.config.configurable import Configurable
+
+
+class TestConfigurable(Configurable):
```
Comment (minrk) [inline tests/docs/test_module.py]: Since this isn't a test module, maybe we should use `sample` naming instead of `test` naming, to avoid confusing pytest's discovery.

File: tests/README.md [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+cd docs
+# --watch:     we watch python files influencing the built documentation
+# --pre-build: we rebuild from scratch as changes to the python files
+#              can influence all built html
+sphinx-autobuild \
+    --open-browser \
+    --watch="test_module.py" \
+    --watch="../../autodoc_traits.py" \
+    --pre-build="rm -rf build" \
+    --ignore
```
Comment (minrk) [inline tests/README.md]: `--ignore` line wrong? It's missing a `\` continuation, but if I add it the command doesn't work. It appears to work if I remove the `--ignore`.

File: tests/docs/source/autoclass/undoc_members.rst [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+=========================
+
+The ``members`` and ``undoc-members`` option combined should present all class
+members, including the traits. And with autodoc_traits installed, the attributes
+should be presented with their help strings as docstrings.
+
+.. autoclass:: test_module.TestConfigurable
+   :noindex:
+   :members:
+   :undoc-members:
```
Comment (minrk) [inline tests/docs/source/autoclass/undoc_members.rst:11]: It doesn't seem quite right to me that adding undocumented members brings in _documented_ traits that weren't there before. It seems like if `members` only picks up non-traits by default, `undoc-members` should _also_ only pick up non-traits by default. If `undoc-members` brings in traits, then `:members:` should bring in _documented_ traits.

File: tests/docs/source/conf.py [hunk 5]
```diff
@@ -0,0 +1,14 @@
+import os
+import sys
+
+project = "autodoc_traits tests"
+extensions = ["autodoc_traits"]
```
Comment (minrk) [inline tests/docs/source/conf.py]: adding napoleon should fix issues with reading formatted docstrings, since those packages use numpy or google-format docstrings.

```suggestion
extensions = [
    "autodoc_traits",
    # sphinx.ext.napoleon is added to avoid warnings if testing with :inherited-members:`
    # where the traitlets base classes members has numpy or google-format docstrings
    "sphinx.ext.napoleon"
]
```

File: tests/docs/source/autoconfigurable/inherited_members.rst [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+TestConfigurable to show up as well as traits from TestConfigurableSubclass.
+
+.. note::
+   Several members from the ``traitlets.HasTraits`` base class are excluded for
+   now as they cause the sphinx-build warning of ``Unexpected section title`` in
+   sphinx 4.
+
+.. autoconfigurable:: test_module.TestConfigurableSubclass
+   :noindex:
+   :inherited-members:
```
Comment (minrk) [inline tests/docs/source/autoconfigurable/inherited_members.rst:9]: I tried to set `:inherited-members: traitlets.HasTraits`, since the [documentation](https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html) suggests this should exclude methods inherited from this class, but it doesn't seem to work.

I've tried it on `autoclass` without autodoc-traits, and `:inherited-members: parent.Class` seems to simply not do what it says.

File: tests/docs/source/autoconfigurable/no_members.rst [hunk 7]
```diff
@@ -0,0 +1,10 @@
+autoconfigurable - no members
+=============================
+
+Without the ``members`` option, one may expect no members to show up, but we
+present them all still, see `Issue27`_.
```
Comment (minrk) [inline tests/docs/source/autoconfigurable/no_members.rst]: I think it's a valid goal to present the option to export only the configurable options, and not the Python API of a class. I'm not saying `autoconfigurable` without members is necessarily the right signature, but that output should be an option.

File: tests/docs/source/index.rst [hunk 8]
```diff
@@ -0,0 +1,9 @@
+autodoc_traits tests
+====================
+
+.. toctree ref: https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-toctree
+.. toctree::
+   :maxdepth: 2
+   :glob:
+
+   */*
```
Comment (minrk) [inline tests/docs/source/index.rst:9]: I did not know about toctree globs, this is so useful!

**Review by minrk — APPROVED** [review]
This is truly wonderful, thanks for putting in all the effort!

Looking through the examples, I think we want to be able to support these cases:

1. document all config options on the class, but no Python and no inheritance (I think this is what `autoconfigurable` without members does now)
2. document all config options _including inherited config_, still without Python APIs (I think this is infeasible because `:inherited-members:` _also_ pulls in Python members, so you'd have to identify and manually exclude all non-configurable members, right?

Maybe the answer is to not use `:inherit-members:`, and instead define a new option like `:inherit-config-members:`. I'm not sure about needing a `:config-members:` equivalent to `:members:`, because `autoconfigurable` doesn't make any sense if you aren't going to document any configurable traits, since that's just the same output as `autoclass`, right?

I don't think you need to address these things here, because what you have is already a massive improvement. I think having these wonderful examples and illustrations merely _reveals_ that these are cases we should think about. So maybe it should be an Issue. Then again, nobody's asked for the inherit-config-but-not-python case, so maybe it doesn't matter.

**Comment by minrk** [comment]
Actually, now that I think about it, the inherited-members being backwards means we've always been showing _inherited_ config and never showing just on-class. This is the case in jupyterhub, because [log_level is inherited from base Application](https://jupyterhub.readthedocs.io/en/stable/api/app.html#jupyterhub.app.JupyterHub.log_level) and is documented. I think that's probably what folks usually want, so maybe it is important to keep it working, and maybe even still be the default.

What do you think is the best way to signal that?

**Comment by minrk** [comment]
Sorry for mixed messages - that doesn't mean I think it needs to block this PR, but I think it _does_ mean we need to open an Issue about it that blocks _release_ so we can make an explicit decision about it.

File: tests/docs/source/autoconfigurable/no_members.rst (hunk 7 above)
Comment (consideRatio) [reply tests/docs/source/autoconfigurable/no_members.rst]: I agree, I commented in https://github.com/jupyterhub/autodoc-traits/issues/27#issuecomment-1351480422 about it.

**Review by minrk — APPROVED** [review]

File: tests/docs/source/autoconfigurable/inherited_members.rst (hunk 6 above)
Comment (consideRatio) [reply tests/docs/source/autoconfigurable/inherited_members.rst:9]: I recall reading upstream sphinx source code and noting its used as a boolean flag rather than a list of members.

File: tests/docs/source/autoclass/undoc_members.rst (hunk 4 above)
Comment (consideRatio) [reply tests/docs/source/autoclass/undoc_members.rst:11]: Hmmm hmm but this is the `autoclass` trait.

`autoclass` will discard all members that are considered undocumented, and attributes are considered that. We rely on a hack in `autoconfigurable` to make traitlets look documented to the sphinx.ext.autodoc logic (which is inspecting `__doc__` to filter out members picked up, but considered undocumented).

Overall, I think we should not bother with the `autoclass` directive. We influence it only slightly sometimes, if its used on a class with traitlets and those traitlets are presented. Everybody should be using `autoconfigurable` for those classes.

If possible, I'd like autoclass to consider the traits with help strings documented members, but that doesn't seem possible without overriding the directive. I was trying to do that initially, but it led to sphinx warnings etc.

**Comment by consideRatio** [comment]
@minrk I've addressed all comments, either by adding commits or commenting. On my end, this is ready for a merge but are pinging you to re-review added commits.

File: tests/docs/source/autoconfigurable/inherited_members.rst (hunk 6 above)
Comment (consideRatio) [reply tests/docs/source/autoconfigurable/inherited_members.rst:9]: HMmm, what version of sphinx was used? See https://github.com/sphinx-doc/sphinx/pull/10325 added for sphinx 5.0

File: tests/docs/source/autoclass/undoc_members.rst (hunk 4 above)
Comment (minrk) [reply tests/docs/source/autoclass/undoc_members.rst:11]: Ah, I understand. Let's consider that a bug that we don't need to address any time soon, and open a low-priority issue for it (#30).

File: tests/docs/source/autoconfigurable/inherited_members.rst (hunk 6 above)
Comment (minrk) [reply tests/docs/source/autoconfigurable/inherited_members.rst:9]: 5.3.0. Not our problem, for now. It seems like it will be desirable _pretty often_ for documenting inheritors of HasTraits, which has _a lot_ of methods.

File: tests/docs/source/autoconfigurable/specified_members.rst [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+
+In this test we list the member ``method`` and ``trait_nohelp``. One may expect
+no members besides these to show up, but we present them and all trait members
+still, see `Issue27`_.
+
+.. _Issue27: https://github.com/jupyterhub/autodoc-traits/issues/27
+
+.. autoconfigurable:: sample_module.SampleConfigurable
+   :noindex:
+   :members: method,trait_nohelp
```
Comment (minrk) [inline tests/docs/source/autoconfigurable/specified_members.rst:15]: `trait_nohelp` is not included. Does the test cover this?

Screenshot 2022-12-15 at 12 54 26

Comment (consideRatio) [reply tests/docs/source/autoconfigurable/specified_members.rst:15]: Hmmm... It is for me.

Here is the segment of code testing that it is: https://github.com/consideRatio/autodoc-traits/blob/pr/refresh-and-add-tests/tests/test_autodoc_traits.py#L76-L85

Here is a screenshot from my computer using sphinx 4, tested to do the same with sphinx 5 and sphinx 6beta3 also: 
image

Comment (minrk) [reply tests/docs/source/autoconfigurable/specified_members.rst:15]: Huh, ok. Must be something weird in my setup.

**Comment by minrk** [comment]
This is terrific, great work!

**Comment by consideRatio** [comment]
Thanks for reviewing this @minrk!!!
